### PR TITLE
Options as destructured parameters (and sortFn for static functions, while we're at it)

### LINKED ISSUE #219: Add the sortFn to toMultihash, fromMultihash, etc
Currently we can pass a custom `sortFn` to the `Log` constructor. The same capability needs to be applied to the `fromMultihash` functions and similar

### PR DESCRIPTION
**Pull request by aphelionz**
## Proposal

Migrate from this:
```javascript
static async fromCID (ipfs, access, identity, cid, length = -1, exclude, onProgressCallback)
```

to this:
```javascript
static async fromCID (ipfs, access, identity, cid, { length = -1, exclude, onProgressCallback, sortFn })
```

For the following  functions
- [x] `Log` constructor
- [x] `Log.fromJSON`
- [x] `Log.fromEntry`
- [x] `Log.fromCID`
- [x] `Log.fromMultihash`
- [x] `Log.fromEntryCID`
- [x] `Log.fromEntryHash`
- [x] `LogIO.fromCID`
- [x] `LogIO.fromEntry`
- [x] `LogIO.fromEntryCID`
- [x] `EntryIO.fetchParallel`
- [x] `EntryIO.fetchAll`

This is a nice middle ground between signature transparency and  a more convenient developer experience.

Resolves #219 

## Other changes (e.g. bug fixes, UI tweaks, refactors)

I also added the `sortFn` argument to the destructured options, since it was quick and easy to do. Need tests.

## TODO

- [x] Complete refactor of all functions in the description above 
- [x] Figure out what to do with standardjs `  /home/mark/Projects/orbitdb/ipfs-log/src/log.js:467:5: Return statement should not contain assignment.`
- [x] Commit build output files

## TODO: Docstring Updates
- [x] `Log` constructor
- [x] `Log.fromJSON`
- [x] `Log.fromEntry`
- [x] `Log.fromCID`
- [x] `Log.fromMultihash`
- [x] `Log.fromEntryCID`
- [x] `Log.fromEntryHash`
- [x] `LogIO.fromCID`
- [x] `LogIO.fromEntry`
- [x] `LogIO.fromEntryCID`
- [x] `EntryIO.fetchParallel`
- [x] `EntryIO.fetchAll`

## TODO: Tests

- [x] `Log.fromEntry`
- [x] `Log.fromCID`
- [x] `Log.fromMultihash`
- [x] `Log.fromJSON`
- [x] `Log.fromEntryCID`
- [x] `Log.fromEntryHash`

### CODE REVIEW

**Comment by aphelionz** [comment]
cc @satazor

**Comment by aphelionz** [comment]
Here's an interesting artifact of moving to this approach. Consider the simple example

```javascript
 static async fromMultihash (ipfs, access, identity, multihash,
   { length = -1, exclude, onProgressCallback, sortFn } = {}) {
   return Log.fromCID(ipfs, access, identity, multihash,       
     { length, exclude, onProgressCallback, sortFn } = {})
 }               
```

When linting, I get this error: `Return statement should not contain assignment.` But it's not entirely intuitive how to fix that. Default values for `fromMultihash` but not for `fromCID`? or to recreate the object then pass it in to `fromCID`?

Open for discussion.

**Comment by satazor** [comment]
@aphelionz 

- Shouldn't the Log constructor also be changed to have `options` instead of all the optional arguments it has now? I would also look for other functions that have the same issue and change as well.
- If we want to make this backwards-compatible, we should create a helper that would take all the optional arguments and their order and normalize it to an options object. This way, we could reuse it everywhere and maintain compatibility.

```js
static async fromCID (ipfs, access, identity, cid, ...rest) {
    const { length, exclude, onProgressCallback } = normalizeOptions(rest, ['length', 'exclude', 'onProgressCallback'])
}
```

**Comment by aphelionz** [comment]
@satazor Yes, I thought of that and others - such as the `Entry` constructor that have similar signatures. However, I'll admit I'm just nervous about making such a big change.

`normalizeOptions` seems like a good idea as I love the idea of not having to change the tests and break the API. So, that seems pretty good.

@shamb0t I would love another set of eyes on this as well if you have a second.

**Comment by aphelionz** [comment]
@satazor The matter of default values comes into question with the `normalizeOptions` approach. Currently we have, for example, `length: -1`. I'm not sure how to do that effectively here. perhaps an object merge after `normalizeOptions` is called?

**Comment by shamb0t** [comment]
This looks good so far @aphelionz! It would be preferable to change all function signatures to match this approach. Agreed that we can have a defaultOptions object and just merge them with `Object.assign({}, defaultOptions, normalizedOptions)` after normalizing. We should also add `access` to the options as there is a default value. Re: normalize helper function, I think it may be preferable to break the API (given the next release will do that anyways) I like the idea of a "normalizeOptions' method that can sanitize and ensure correct shape but I dont think backwards compatibility is necessary here so no need to remember the order of arguments passed.

**Comment by aphelionz** [comment]
If we're breaking the API, then it's probably fine and simpler to keep it as the destructured arguments in te function signature instead of the spread syntax.

So, keep:
```javascript
static async fromCID (ipfs, access, identity, cid, { length = -1, exclude, onProgressCallback, sortFn })
```

instead of:
```javascript
static async fromCID (ipfs, access, identity, cid, ...options)
```

This handles defaults, is explicit in what it expects, and doesn't require the normalize help function.

If we agree, then all I need is a list of functions to give this treatment to (`Log` constructor, `Entry` constructor) and then I can move forward.

**Comment by shamb0t** [comment]
@aphelionz I agree that way is simpler.

Looks to me like the functions that need treatment are the `constructor` in Log,  `static async fromJSON`, `static async fromEntry`, `static async fromEntryCid`, `static async fromMultihash` in LogIO and `static async fetchParallel` and `static async fetchAll` in EntryIO.

**Comment by satazor** [comment]
> If we're breaking the API, then it's just simpler to keep it as the destructured arguments in the function signature instead of the spread syntax.

I agree, if we are doing a breaking change then keep it simple. If we want to maintain compatibility, create the `normalizeOptions` function.

**Comment by satazor** [comment]
@aphelionz This looks good to me!

**Comment by aphelionz** [comment]
Getting a strange error on build. It looks familiar but I forget how we solved it in the past.

```
ERROR in ./node_modules/datastore-level/node_modules/level-js/iterator.js
Module not found: Error: Can't resolve 'idb-readable-stream' in '/home/mark/Projects/orbitdb/ipfs-log/node_modules/datastore-level/node_modules/level-js'
 @ ./node_modules/datastore-level/node_modules/level-js/iterator.js 4:24-54
 @ ./node_modules/datastore-level/node_modules/level-js/index.js
 @ ./node_modules/datastore-level/src/index.js
 @ ./node_modules/ipfs/node_modules/ipfs-repo/src/default-options-browser.js
 @ ./node_modules/ipfs/node_modules/ipfs-repo/src/index.js
 @ ./node_modules/ipfs/src/core/boot.js
 @ ./node_modules/ipfs/src/core/index.js
 @ ./examples/browser/index.js
```

cc @shamb0t @satazor @haadcode

**Comment by aphelionz** [comment]
Fixed the above error

**Comment by aphelionz** [comment]
Here's another interesting finding.

There we were no tests written for `fromJSON` so I added one. This uncovered a bug,  a `log.length` discrepancy where it would return a different value than expected.

This comes from `fetchParallelit`, which traces all traversal paths from heads concurrently. For example, one traversal path would be A1 - A10, while another would be A1, B1, A2, B2, A3, etc... This is expected and good. However, but when it merges them it does not filter out duplicates. In the case of this test, 21 entries were returned instead of the expected 16.

You can [see here](https://github.com/orbitdb/ipfs-log/pull/220/commits/d611d96134fb1df7a35a7153127563ec7312577a#diff-053e0c19792f28d21a88aa2412aefcb1R25) where I augmented the `concatArrays` function to be the `uniquelyConcat` arrays function.

**Comment by aphelionz** [comment]
@satazor @shamb0t @haadcode This is ready for a comprehensive review and potential merge.

File: src/entry-io.js [hunk 1]
```diff
@@ -9,34 +9,48 @@ class EntryIO {
...
+    const getHashes = e => e.hash
+    const uniquelyConcatArrays = (arr1, arr2) => {
+      // Add any new entries to arr1
+      const hashes = arr1.map(getHashes)
+      arr2.forEach(entry => {
+        if (hashes.indexOf(entry.hash) === -1) arr1.push(entry)
+      })
+      return arr1
+    }
+    const flatten = (arr) => arr.reduce(uniquelyConcatArrays, [])
```
Comment (haadcode) [inline src/entry-io.js]: @aphelionz would like to understand why the change here? It seems expensive to go through the arrays every time to get the hashes plus checking the uniqueness. Iirc, `fetchAll` should take care of this in that it only returns uniques.

File: test/log.spec.js [hunk 2]
```diff
@@ -535,6 +562,35 @@ Object.keys(testAPIs).forEach((IPFS) => {
         })
       })
 
+      describe('fromEntryHash', async () => {
+        afterEach(() => {
+          if (Log.fromCID.restore) {
+            Log.fromCID.restore()
```
Comment (shamb0t) [inline test/log.spec.js]: Should this be ```if (Log.fromEntryCid.restore) Log.fromEntryCid.restore()```?

File: test/log.spec.js [hunk 3]
```diff
@@ -394,7 +401,7 @@ Object.keys(testAPIs).forEach((IPFS) => {
             id: 'X',
             heads: ['zdpuB2NAQ7cSh9MAfY91QC6Va56pQMJBXBaLoS6uQ1qNxqija']
           }
-          let log = new Log(ipfs, testACL, testIdentity, 'X')
+          let log = new Log(ipfs, testACL, testIdentity, { logId: 'X' })
           await log.append('one')
           const multihash = await log.toMultihash()
           const res = await Log.fromCID(ipfs, testACL, testIdentity, multihash, -1)
```
Comment (shamb0t) [inline test/log.spec.js]: Should the `-1` after multihash be in the passed as `{ length: -1 }`?

File: src/entry-io.js (hunk 1 above)
Comment (haadcode) [reply src/entry-io.js]: Understood, thanks. We can take a closer look at the performance optimizations later 👍

File: test/log.spec.js (hunk 3 above)
Comment (aphelionz) [reply test/log.spec.js]: Fixed!

File: test/log.spec.js (hunk 2 above)
Comment (aphelionz) [reply test/log.spec.js]: Fixed

**Comment by aphelionz** [comment]
@haadcode @shamb0t @satazor Shall we merge?

**Comment by shamb0t** [comment]
@aphelionz LGTM :+1:
